File: apps/ios/TaakHelden/Core/DesignSystem/CelebrationService.swift
import Foundation
import UIKit

@Observable
final class CelebrationService {
    private let sound = SoundEffectService()
    private(set) var confettiToken = 0

    func celebrateTaskCompleted(reduceMotion: Bool) {
        // Haptic and sound celebrate the completion regardless of motion preference.
        let generator = UINotificationFeedbackGenerator()
        generator.notificationOccurred(.success)

        if AppSettings.childSoundsEnabled {
            sound.playTaskCompleteChime()
        }

        // Only trigger confetti / motion overlay when Reduce Motion is off.
        // ConfettiOverlay already shows a static glow when reduceMotion is true,
        // but we must not increment the token at all so no animation fires.
        if !reduceMotion {
            confettiToken += 1
        }
    }
}

File: apps/android/core/src/main/kotlin/nl/taakhelden/core/celebration/CelebrationService.kt
package nl.taakhelden.core.celebration

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/** Platform hooks for the reward moment (haptics + chime). */
public interface CelebrationEffects {
    public fun playSuccessHaptic()
    public fun playTaskCompleteChime()
    public val childSoundsEnabled: Boolean
}

/** No-op effects for tests and previews. */
public class NoopCelebrationEffects(
    override val childSoundsEnabled: Boolean = true,
) : CelebrationEffects {
    override fun playSuccessHaptic(): Unit = Unit
    override fun playTaskCompleteChime(): Unit = Unit
}

/**
 * Fires the reward moment when a child finishes a task.
 *
 * Haptic and chime play regardless of the motion preference — they are the celebration
 * for someone who cannot or does not want to see animation. The confetti token is only
 * bumped when Reduce Motion is off, so no animation is scheduled at all in that case.
 */
public class CelebrationService(
    private val effects: CelebrationEffects = NoopCelebrationEffects(),
) {
    private val _confettiToken = MutableStateFlow(0)
    public val confettiToken: StateFlow<Int> = _confettiToken.asStateFlow()

    public fun celebrateTaskCompleted(reduceMotion: Boolean) {
        effects.playSuccessHaptic()

        if (effects.childSoundsEnabled) {
            effects.playTaskCompleteChime()
        }

        if (!reduceMotion) {
            _confettiToken.value += 1
        }
    }
}
